dali/image/tiff_libtiff_impl.cc

```diff
@@ -0,0 +1,308 @@
+    toff_t new_pos = helper->buf_pos_;
+    switch (whence) {
+    case SEEK_SET:
+      new_pos = offset;
+      break;
+    case SEEK_CUR:
+      new_pos += offset;
+      break;
+    case SEEK_END:
+      new_pos = size + offset;
+      break;
+    }
+    new_pos = std::min(new_pos, size);
+    helper->buf_pos_ = static_cast<size_t>(new_pos);
+    return new_pos;
+  }
+
+  static int map(thandle_t handle, void **base, toff_t *size) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    auto &buf = helper->buf_;
+    *base = const_cast<uint8_t*>(buf.data());
+    *size = buf.size();
+    return 0;
+  }
+
+  static toff_t size(thandle_t handle) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    return helper->buf_.size();
+  }
+
+  static int close(thandle_t handle) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    delete helper;
+    return 0;
+  }
+};
+
+}  // namespace detail
+
+TiffImage_LibtiffImpl::TiffImage_LibtiffImpl(const uint8_t *encoded_buffer, size_t length, DALIImageType image_type)
+    : GenericImage(encoded_buffer, length, image_type)
+    , buf_({encoded_buffer, static_cast<ptrdiff_t>(length)})
+    , buf_pos_(0) {
+  tif_.reset(
+    TIFFClientOpen("", "r",
+                   reinterpret_cast<thandle_t>(
+                     new detail::BufDecoderHelper(buf_, buf_pos_)),
+                   &detail::BufDecoderHelper::read,
+                   &detail::BufDecoderHelper::write,
+                   &detail::BufDecoderHelper::seek,
+                   &detail::BufDecoderHelper::close,
+                   &detail::BufDecoderHelper::size,
+                   &detail::BufDecoderHelper::map,
+                   /*unmap=*/0));
+
+  LIBTIFF_CALL(
+    TIFFGetField(tif_.get(), TIFFTAG_IMAGELENGTH, &shape_[0]));
+  LIBTIFF_CALL(
+    TIFFGetField(tif_.get(), TIFFTAG_IMAGEWIDTH, &shape_[1]));
+  LIBTIFF_CALL(
+    TIFFGetField(tif_.get(), TIFFTAG_SAMPLESPERPIXEL, &shape_[2]));
+  is_tiled_ = static_cast<bool>(
+    TIFFIsTiled(tif_.get()));
+  LIBTIFF_CALL(
+    TIFFGetField(tif_.get(), TIFFTAG_BITSPERSAMPLE, &bit_depth_));
+  DALI_ENFORCE(bit_depth_ <= 64,
+    "Unexpected bit depth: " + std::to_string(bit_depth_));
+
+  // optional
+  TIFFGetField(tif_.get(), TIFFTAG_ORIENTATION, &orientation_);
+}
+
+Image::ImageDims TiffImage_LibtiffImpl::PeekDims(const uint8_t *encoded_buffer, size_t length) const {
+  DALI_ENFORCE(encoded_buffer != nullptr);
+  assert(encoded_buffer == buf_.data());
+  return std::make_tuple(static_cast<size_t>(shape_[0]),
+                         static_cast<size_t>(shape_[1]),
+                         static_cast<size_t>(shape_[2]));
+}
+
+std::pair<std::shared_ptr<uint8_t>, Image::ImageDims>
+TiffImage_LibtiffImpl::DecodeImpl(DALIImageType image_type, const uint8 *encoded_buffer, size_t length) const {
+  if (!CanDecode(image_type)) {
+    DALI_WARN("Warning: Falling back to GenericImage");
```

Comment:
Do you have the offending image name/path to ad to this warning?

Reply:
Nope, `Image` only decodes, it's agnostic of reading...

Reply:
unfortunately not